src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -865,6 +876,70 @@ async Task<IEnumerable<IAccount>> ITokenCacheInternal.GetAccountsAsync(Authentic
             return accounts;
         }
 
+        async Task<IDictionary<string, ITenantProfile>> ITokenCacheInternal.GetTenantProfilesAsync(AuthenticationRequestParameters requestParameters, string homeAccountId)
+        {
```

Comment:
this is very similar to GetAccountsAsync(). do we need the ADAL part and to do instance discovery? can the shared code be moved to another method? i would think we would have already done instance dis. do you know the perf impact?

Reply:
if you can pick up the instance dis info elsewhere, this won't need to be async.

Reply:
Instance discovery part cannot be moved elsewhere but by the time we reach here instance discovery is already done and cached. Since we are querying the cached instance discovery there will be no perf impact.

Reply:
I modified this. Neha has a good point that instance discovery needs to be performed, otherwise we risk not being able to load some id tokens. 

However, there is no need for ADAL legacy at all. Only RTs from ADAL are usable. Tenant Profiles are a nice view over MSAL ID tokens.